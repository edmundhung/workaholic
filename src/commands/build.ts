import { Command } from 'commander';
import * as esbuild from 'esbuild';
import fs from 'fs';
import path from 'path';
import type { Build, Entry, Options } from '../types';
import { getRelativePath, getWranglerConfig, getWranglerDirectory } from '../utils';

interface BuildOptions {
  source: string;
  config?: string;
}

async function parseFile(root: string, filePath: string): Promise<Entry> {
  const value = await fs.promises.readFile(filePath);
  const key = getRelativePath(root, filePath);

  return {
    key,
    value: value.buffer,
  };
}

async function parseDirectory(directory: string): Promise<string[]> {
  let list: string[] = [];

  for (const dirent of await fs.promises.readdir(directory, { withFileTypes: true })) {
    if (dirent.isDirectory()) {
      const paths = await parseDirectory(`${directory}/${dirent.name}`);

      list.push(...paths);
    } else if (dirent.isFile()) {
      list.push(`${directory}/${dirent.name}`);
    }
  }

  return list;
}

function assignNamespace(namespace: string, entries: Entry[]): Entry[] {
  return entries.map<Entry>(entry => ({ ...entry, key: `${namespace}://${entry.key}` }));
}

export default async function build({ source, config = path.resolve(__dirname, '../defaultConfig') }: BuildOptions): Promise<Entry[]> {
  const files = await parseDirectory(source);
  const entries = await Promise.all(files.map(file => parseFile(source, file)));
  const build = initialiseBuild(config);
  const result = await Promise.resolve(build(entries));

  return Object.entries(result).flatMap(([namespace, entries]) => assignNamespace(namespace, entries));
}

async function resolveConfig(root: string, config: string): Promise<string> {
  const target = path.resolve(root, './node_modules/.workaholic/', config);

  await esbuild.build({
    entryPoints: [config.startsWith('.') ? path.resolve(root, config) : config],
    outfile: target,
    format: 'cjs',
    target: 'node12',
  });

  return target;
}

function initialiseBuild(config: string): Build {
  return require(config).setupBuild();
}

function stringifyEntries(entries: Entry[]): string {
  const result = entries.map(entry => {
    if (typeof entry.value === 'string') {
      return entry;
    }

    return {
      ...entry,
      value: Buffer.from(entry.value).toString('base64'),
      base64: true,
    };
  });

  return JSON.stringify(result, null, 2);
}

export function makeBuildCommand(): Command {
  const command = new Command('build');

  command
    .description('Build worker kv data')
    .argument('<output>', 'output file path')
    .option('--source [path]', 'path of the source directory')
    .action(async (output, options) => {
      const cwd = process.cwd();
      const root = await getWranglerDirectory();
      const config = await getWranglerConfig(root);
      const workaholic = config.getWorkaholicOptions();
      const entries = await build({
        source: path.resolve(root, options.source ? path.resolve(cwd, options.source) : workaholic.source),
        config: await resolveConfig(root, workaholic.config),
      });

      fs.promises.writeFile(path.resolve(cwd, output), stringifyEntries(entries));
    });

  return command;
}
