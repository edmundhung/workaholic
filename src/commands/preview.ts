import * as fs from 'fs/promises';
import { Miniflare } from 'miniflare';
import rimraf from 'rimraf';

async function preview(source: string, binding: string): Promise<void> {
  console.log('[workaholic] Empty ./.mf/kv');
  rimraf.sync('./.mf/kv');
  console.log('[workaholic] Cleanup done');

  console.log('[workaholic] Persisting KV on Miniflare');

  const mf = new Miniflare({
    script: `addEventListener("fetch", () => {});`,
    buildCommand: '',
    kvPersist: true,
  });

  const content = await fs.readFile(source, 'utf-8');
  const namespace = await mf.getKVNamespace(binding);

  const entries = JSON.parse(content);

  if (!Array.isArray(entries)) {
    throw new Error('[workaholic] source json must be an array');
  }

  await Promise.all(
    entries.map(entry =>
      namespace.put(entry.key, entry.value, { metadata: entry.metadata }),
    ),
  );
  console.log('[workaholic] KV persisted on Miniflare');
}


export const command = 'preview <data>';

export const describe = 'Persist data on miniflare';

export const builder = {
  data: {
    describe: 'data path',
    type: 'string'
  },
  binding: {
    describe: 'wrangler binding name',
    type: 'string',
    default: 'workaholic'
  }
};

export async function handler(argv) {
  await preview(argv.data, argv.binding);
}
