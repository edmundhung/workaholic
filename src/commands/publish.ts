import { Command } from 'commander';
import fetch, { Response } from 'node-fetch';
import path from 'path';
import { Entry } from '../types';
import { parseData, getWranglerConfig, getWranglerDirectory } from '../utils';

export async function publish(entries: Entry[], { accountId, namespaceId, token }: { accountId: string, namespaceId: string, token: string }): Promise<Response> {
  return fetch(
    `https://api.cloudflare.com/client/v4/accounts/${accountId}/storage/kv/namespaces/${namespaceId}/bulk`,
    {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify(entries),
    },
  );
}

export default function makePublishCommand(): Command {
  const command = new Command('publish');

  command
    .description('Publish kv data to Cloudflare')
    .argument('<source>', 'source file path')
    .option('--preview', 'publish to the preview namespace instead of production', false)
    .action(async (source, options) => {
      const entries = await parseData(path.resolve(process.cwd(), source));
      console.log('[workaholic] Reading config from wrangler.toml');
      const root = await getWranglerDirectory();
      const config = await getWranglerConfig(root);
      const workaholic = config.getWorkaholicOptions();
      const accountId = config.getAccountId();
      const namespaceId = config.getNamespaceId(workaholic.binding, options.preview);
      const token = process.env.CF_API_TOKEN;

      if (!namespaceId) {
        console.log(`[workaholic] Environemnt variable CF_API_TOKEN is missing;`)
        return;
      }

      if (!token) {
        console.log(`[workaholic] Environemnt variable CF_API_TOKEN is missing;`)
        return;
      }

      console.log(`[workaholic] Updating KV with binding "${workaholic.binding}" for account "${accountId}" and namespace "${namespaceId}"`);
      const response = await publish(entries, { accountId, namespaceId, token });
      const result = await response.text();
      console.log(`[workaholic] Update finish with status ${response.status} and result ${result}`);
    });

  return command;
}
