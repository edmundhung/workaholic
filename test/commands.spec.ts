import { Miniflare } from 'miniflare';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import path from 'path';
import build from '../src/commands/build';
import preview from '../src/commands/preview';
import publish from '../src/commands/publish';
import type { Entry } from '../src/types';

function encodeEntry(entry: Entry, encoding: string): Entry {
  return {
    ...entry,
    value: Buffer.from(entry.value).toString(encoding),
    base64: encoding === 'base64',
  };
}

describe('commands', () => {
  const server = setupServer();

  beforeAll(() => {
    server.listen();
  });

  afterEach(() => {
    server.resetHandlers();
  })

  afterAll(() => {
    server.close();
  });

  it('supports building entries based on the fixtures', async () => {
    const entries = await build({ source: path.resolve(__dirname, './fixtures') });

    expect(entries.map(entry => encodeEntry(entry, 'utf-8'))).toMatchSnapshot();
  });

  it('supports previewing entries on miniflare', async () => {
    const mf = new Miniflare({
      script: `addEventListener("fetch", () => {});`,
      buildCommand: '',
    });
    const fixtures = await build({ source: path.resolve(__dirname, './fixtures') });
    const namespace = await preview(mf, 'test', fixtures);

    for (const entry of fixtures) {
      expect(await namespace.getWithMetadata(entry.key)).toEqual({
        value: Buffer.from(entry.value).toString('utf-8'),
        metadata: entry.metadata ?? null,
      });
    }
  });

  it('supports publishing entries to cloudflare', async () => {
    const accountId = 'foo';
    const namespaceId = 'bar';
    const token = 'test-token';
    const entries = await build({ source: path.resolve(__dirname, './fixtures') });
    const fixtures = entries.map(entry => encodeEntry(entry, 'base64'));

    server.use(
      rest.put('https://api.cloudflare.com/client/v4/accounts/:accountId/storage/kv/namespaces/:namespaceId/bulk', (req, res, ctx) => {
        const { accountId, namespaceId } = req.params

        if (req.params.accountId !== accountId || req.params.namespaceId !== namespaceId || req.headers.get('Authorization') !== `Bearer ${token}`) {
          return res(
            ctx.status(400),
          );
        }

        return res(
          ctx.status(200),
          ctx.json(req.body),
        );
      }),
    );
    const response = await publish(fixtures, { accountId, namespaceId, token });

    expect(response.status).toBe(200);
    expect(await response.json()).toEqual(fixtures);
  });
});
