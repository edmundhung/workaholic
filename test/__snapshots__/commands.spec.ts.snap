// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands supports building entries based on the fixtures 1`] = `
Array [
  Object {
    "base64": false,
    "key": "assets://foo/de-hostis-habetur.md",
    "value": "---
title: De hostis habetur
description: Lorem markdownum Laurentes tangor imagine, pallet lacertos ipsius cornua amantibus cum non taceam, post stagnata
---

# De hostis habetur

## Gladii et in celebrant

Lorem markdownum *Laurentes* tangor imagine, pallet lacertos ipsius cornua
amantibus cum non taceam, post stagnata! Sed et noluit. Domo nigris de pietas,
est bis pervenisse oris madebit aper. Recenti iurgia potes avem undis Exigit [an
Lycias](http://sumiuvenis.org/collacogit) anum subita moles tamen **ferrum**
diras.

Condiderant declivis [Byblida](http://tepatentibus.com/regis-est.php) videnda
confidit; cultum [aliqua](http://minervae.com/infirmos.html); mare magni; silvis
ferox haut destituit. In venenis istis infamis dilacerant ligatis Combe domique,
aliter tamen. Suis mihi quamvis suo, ego mox **turbaeque et** manet, sua quam.
Mensis nunc sit erat **Achilles ut** est venatibus animo in possederat nomen,
conceperat coloribus poenam venti *Phrygiis*, quoque.

- A turget hoc vibrata
- Nunc agitant
- Creatas trieterica vulnere Caeneus Sabinis morsu multis
- Minus et repellit
- Subdita transit
- Ego mihi Vulcanum premebat absumere ceditis conspecta

## Siccaeque patria refluitque solidis mediis

Veloxque et putant, Cipe super, spernimus meos urbes, sit sic ventos altera, cum
quati suo. Vixque medeare sic super: fortibus odiis ignota *Ioles*, tuorum
campis. Phrygiae liquido natum. Ubi belli; ripas et tanta latices currere
strinxit metus Agamemnonis dixit Cecropide atque recens sumit dicentem. Pendere
mota.

Hostem sed fonti et nunc **consurgit tempora**, ad [mandarat
calcavit](http://www.consumitis.io/parsquetantum) nivea. In viae arboris unda
**fortibus tempus** instructo, et et inposuit Aegeus. Cupidoque flores; sacri
digna fingam, flos taurum, ipse?

    serp = system_processor_inkjet(rfid + cycle_secondary_bus, defaultPop(data,
            key, 2));
    if (3 + forum_camelcase_traceroute(2, pppAssociationBoot,
            switchUdpStreaming) * toolbarThyristorCpl(uddi(search,
            cybersquatter_service_tunneling), node_facebook(
            undo_framework_start, -2))) {
        unicode = wrap - 73 + thick;
        bookmark_pmu_utility += 1;
        power.barcraft_lcd *= 50;
    }
    driverDevice = wiki_data_resolution(ole, kibibyte_memory, jspMotherboard) +
            dvd_row;
    if (copy(task_unit_multicasting, clock(2, non, mirrored))) {
        visualSpoolingAtm.php_offline_port = navigationPasteBlu;
    } else {
        batch_jpeg_log(up(extension_donationware_array), fios(edutainment,
                orientation_cifs_spooling, scroll_samba_northbridge), -2);
    }

Amatas senectus sollertia et alioque dea, iussos, petit Pelia, ornat. Opus
*Hecate* sive satis, siquis *membra vellem virginitas* omnia dantibus. Hunc
movente suo silentia!
",
  },
  Object {
    "base64": false,
    "key": "assets://foo/mulcet-vincere.md",
    "value": "---
title: Mulcet vincere
description: Lorem markdownum gentemque incerti, genibus? In Pelion est
---

# Mulcet vincere

## Vel longoque coniugialia erat

Lorem markdownum gentemque incerti, genibus? In Pelion est; seu et tota nomenque
corpora tangunt Haemoniae cupidine per, *et silvas* vidisse redibam, et. Femina
cornu, inmurmurat vetantis imagine dicit, ope superesse **consensu desertaque
parte** retemptat *inmitis*, omnia Polyxena.

1. Lacrimas in dicta saepius ista mutua nomen
2. Mors Romam reges revocamina decebat veluti
3. Mutavit iaculante portat posuit
4. Erat mei celebrabant densa mansit

## Quam mihi diversa

Dicenti lacrimis mea genero tamen divino mandatam rurigenae provida. Obruit
constitit et tenebrosa vestrae *comitum talia miserrima* generosos faciem nocti
annua. Erit modo, secura: sit undis aglauros ponit illa, Ichnobatesque tenuisse.
Interceperit silvas, sua nymphas solis ab **fiant**.

- Quo quibus
- Saevitiam utque procul huius
- Sua certe palla stipes cervo et incepto
- Undas barbam tu data cornu

Det arbiter in bracchia quae. Sed Iuppiter facta insignia, more vires celebres
difficilis, et velint **frustra**. Quinque relictum. Puto omnes; tamen Scyrum
debuerant nubila ture virtute lacrimis lustro, e ictibus illa pependit secundis
opem collumque videbatur.

## Tigres virga falleris iugulo fuissem at exitio

Ulixe micabant excussaque mittam! Ab liquit, reddit saepe quem domo ardor,
segetes nec hunc fixurus et tristis in praemia. Signum en pressant, sic vices
nec infamia morsus et iacentia.

Nares placere et suarum voluntas tremensque coniugialia fecit, fuit Iove quoniam
valido nec est quoque amari, abesse. Nec et tenui, fuerat tecto querellae sui.
Fugit monilia Caesaris sanguine: hic
[nec](http://ulteriusinflata.io/omnibus-cum) annos.

1. Illic pulchram novique fiducia lapsa
2. Tyrioque spina genitoris nebulasque vertice invictos sonat
3. Cytherea mentis incubuitque sola ingenti

Tria spectari terram et caeruleo erit Nebrophonosque natura quo medias. *Amens*
diu nisi, sepulcro et longius eminet annos absit formas.
",
  },
  Object {
    "base64": false,
    "key": "assets://sample-json.json",
    "value": "{
  \\"text\\": \\"Phasellus finibus nunc quis leo convallis, eu pulvinar augue egestas\\",
  \\"number\\": -2120385380.6364622,
  \\"boolean\\": true,
  \\"list\\": [
    61125266,
    0.0
  ],
  \\"metadata\\": {
    \\"title\\": \\"Lorem Ipsum\\",
    \\"description\\": \\"Lorem markdownum timidumque putat. Ipse **reddit aderat luctus** enim tutam semine Tiberinaque oppositas non adhuc metaque ferant\\"
  },
  \\"arrayOfObject\\": [
    { \\"foo\\": \\"bar\\" }
  ],
  \\"object\\": {
    \\"date\\": \\"1970-01-01T00:00:00.000Z\\",
    \\"list\\": [
      \\"Maecenas fringilla massa non venenatis porta\\"
    ]
  }
}
",
  },
  Object {
    "base64": false,
    "key": "assets://sample-markdown.md",
    "value": "---
title: Lorem Ipsum
description: >-
  Lorem markdownum timidumque putat. Ipse **reddit aderat luctus** enim
  tutam semine Tiberinaque oppositas non adhuc metaque ferant
---

# Mulcet vincere

## Vel longoque coniugialia erat

Lorem markdownum gentemque incerti, genibus? In Pelion est; seu et tota nomenque
corpora tangunt Haemoniae cupidine per, *et silvas* vidisse redibam, et. Femina
cornu, inmurmurat vetantis imagine dicit, ope superesse **consensu desertaque
parte** retemptat *inmitis*, omnia Polyxena.

1. Lacrimas in dicta saepius ista mutua nomen
2. Mors Romam reges revocamina decebat veluti
3. Mutavit iaculante portat posuit
4. Erat mei celebrabant densa mansit

## Quam mihi diversa

Dicenti lacrimis mea genero tamen divino mandatam rurigenae provida. Obruit
constitit et tenebrosa vestrae *comitum talia miserrima* generosos faciem nocti
annua. Erit modo, secura: sit undis aglauros ponit illa, Ichnobatesque tenuisse.
Interceperit silvas, sua nymphas solis ab **fiant**.

- Quo quibus
- Saevitiam utque procul huius
- Sua certe palla stipes cervo et incepto
- Undas barbam tu data cornu

Det arbiter in bracchia quae. Sed Iuppiter facta insignia, more vires celebres
difficilis, et velint **frustra**. Quinque relictum. Puto omnes; tamen Scyrum
debuerant nubila ture virtute lacrimis lustro, e ictibus illa pependit secundis
opem collumque videbatur.

## Tigres virga falleris iugulo fuissem at exitio

Ulixe micabant excussaque mittam! Ab liquit, reddit saepe quem domo ardor,
segetes nec hunc fixurus et tristis in praemia. Signum en pressant, sic vices
nec infamia morsus et iacentia.

Nares placere et suarum voluntas tremensque coniugialia fecit, fuit Iove quoniam
valido nec est quoque amari, abesse. Nec et tenui, fuerat tecto querellae sui.
Fugit monilia Caesaris sanguine: hic
[nec](http://ulteriusinflata.io/omnibus-cum) annos.

1. Illic pulchram novique fiducia lapsa
2. Tyrioque spina genitoris nebulasque vertice invictos sonat
3. Cytherea mentis incubuitque sola ingenti

Tria spectari terram et caeruleo erit Nebrophonosque natura quo medias. *Amens*
diu nisi, sepulcro et longius eminet annos absit formas.
",
  },
  Object {
    "base64": false,
    "key": "assets://sample-toml.toml",
    "value": "text = \\"Phasellus finibus nunc quis leo convallis, eu pulvinar augue egestas\\"
number = -2_120_385_380.6364622
boolean = true
list = [ 61_125_266, 0 ]

[metadata]
title = \\"Lorem Ipsum\\"
description = \\"Lorem markdownum timidumque putat. Ipse **reddit aderat luctus** enim tutam semine Tiberinaque oppositas non adhuc metaque ferant\\"

[[arrayOfObject]]
foo = \\"bar\\"

[object]
date = \\"1970-01-01T00:00:00.000Z\\"
list = [ \\"Maecenas fringilla massa non venenatis porta\\" ]
",
  },
  Object {
    "base64": false,
    "key": "assets://sample-yaml.yaml",
    "value": "text: Phasellus finibus nunc quis leo convallis, eu pulvinar augue egestas
number: -2120385380.6364622
boolean: true
list:
- 61125266
- 0
metadata:
  title: Lorem Ipsum
  description: >-
    Lorem markdownum timidumque putat. Ipse **reddit aderat luctus** enim
    tutam semine Tiberinaque oppositas non adhuc metaque ferant
arrayOfObject:
- foo: bar
object:
  date: '1970-01-01T00:00:00.000Z'
  list:
  - Maecenas fringilla massa non venenatis porta
",
  },
  Object {
    "base64": false,
    "key": "assets://test-config.js",
    "value": "
export let setupBuild = () => {
  return entries => ({
    test: entries,
  });
};

export let setupQuery = () => {
  return query => (namespace, slug) => {
    switch (namespace) {
      case 'foo':
        return query('test', \`foo/\${slug}\`);
      default:
        return query(namespace, slug);
    }
  };
};
",
  },
]
`;
